tags/Coveo/CoveoAnalyticsLogEvent/Coveo Analytics Log Event.tpl

```diff
@@ -0,0 +1,446 @@
+                  }
+                ]
+              },
+              {
+                "type": 3,
+                "mapKey": [
+                  {
+                    "type": 1,
+                    "string": "key"
+                  },
+                  {
+                    "type": 1,
+                    "string": "read"
+                  },
+                  {
+                    "type": 1,
+                    "string": "write"
+                  },
+                  {
+                    "type": 1,
+                    "string": "execute"
+                  }
+                ],
+                "mapValue": [
+                  {
+                    "type": 1,
+                    "string": "coveoua.t"
+                  },
+                  {
+                    "type": 8,
+                    "boolean": true
+                  },
+                  {
+                    "type": 8,
+                    "boolean": true
+                  },
+                  {
+                    "type": 8,
+                    "boolean": false
+                  }
+                ]
+              }
+            ]
+          }
+        }
+      ]
+    },
+    "clientAnnotations": {
+      "isEditedByUser": true
+    },
+    "isRequired": true
+  },
+  {
+    "instance": {
+      "key": {
+        "publicId": "get_referrer",
+        "versionId": "1"
+      },
+      "param": [
+        {
+          "key": "urlParts",
+          "value": {
+            "type": 1,
+            "string": "any"
+          }
+        }
+      ]
+    },
+    "isRequired": true
+  },
+  {
+    "instance": {
+      "key": {
+        "publicId": "read_title",
+        "versionId": "1"
+      },
+      "param": []
+    },
+    "isRequired": true
+  },
+  {
+    "instance": {
+      "key": {
+        "publicId": "get_url",
+        "versionId": "1"
+      },
+      "param": [
+        {
+          "key": "urlParts",
+          "value": {
+            "type": 1,
+            "string": "any"
+          }
+        }
+      ]
+    },
+    "isRequired": true
+  }
+]
+
+
+___SANDBOXED_JS_FOR_WEB_TEMPLATE___
+
+const log = require('logToConsole');
+
+const addToObject = (obj, obj2) => {
+  for (let key in obj2) {
+    if (obj2.hasOwnProperty(key)) {
+      obj[key] = obj2[key];
+    }
+  }
+  return obj;
+};
+
+const getUrl = require("getUrl");
+const getReferrerUrl = require("getReferrerUrl");
+const readTitle = require("readTitle");
+
+const makeTableMap = require('makeTableMap');
+const customDataObject = data.customData ? makeTableMap(data.customData, 'key', 'value') : {};
+
+const eventData = {
+  location: data.location || getUrl(),
+  referrer: data.referrer || getReferrerUrl(),
+  language: data.language,
+  title: data.title || readTitle(),
+  anonymous: data.isAnonymous,
+  customData: customDataObject
+};
+
+if (data.eventType === "view") {
+  addToObject(eventData, {
+    contentIdKey: data.contentIdKey,
+    contentIdValue: data.contentIdValue,
+  });
```

Comment:
Here these if...else are probably unnecessary. You could just have:

```
const eventData = {
  location: data.location || getUrl(),
  referrer: data.referrer || getReferrerUrl(),
  language: data.language,
  title: data.title || readTitle(),
  anonymous: data.isAnonymous,
  customData: customDataObject,
  contentIdKey: data.contentIdKey,
  contentIdValue: data.contentIdValue,
  eventType: data.customEventType,
  eventValue: data.customEventValue
};
```

Because contentIdKey and contentIdValue are conditional based on the selected event, as are customEventType and customEventValue, either pair will always be undefined when the other pair is active. So you can add all key-value pairs to the eventData object, and only the enabled keys will actually be stored.

Just an idea to make it even simpler :)

Reply:
It is simpler, but I like having those `if/else`. 😄 

Those are keys that should only be set with this specific type of event, so I think it better clarifies the intent, just by looking at the code.

Reply:
Understood!